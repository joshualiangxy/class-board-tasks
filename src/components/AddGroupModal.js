import React, { useState } from 'react';
import Modal from 'react-modal';
import { startNewGroup } from '../actions/groups';
import { connect } from 'react-redux';

// TODO: Add the ability to add other users into the group
const AddGroupModal = ({ isOpen, onRequestClose, startNewGroup }) => {
  const [groupName, setGroupName] = useState('');
  const [module, setModule] = useState('');
  const [error, setError] = useState('');

  const onGroupNameChange = e => setGroupName(e.target.value);

  const onModuleChange = e => {
    const module = e.target.value.toUpperCase();

    setModule(module);
  };

  const onCancel = () => {
    setError('');
    setGroupName('');
    setModule('');
    onRequestClose();
  };

  const onSubmit = e => {
    e.preventDefault();
    const submittedName = groupName.trim();
    const submittedMod = module.trim();
    if (!submittedName) {
      setError('Please enter a group name');
    } else {
      setError('');
      setGroupName('');
      setModule('');
      startNewGroup(submittedName, submittedMod).then(() => onRequestClose());
    }
  };

  return (
    <Modal
      isOpen={isOpen}
      contentLabel="New Group"
      onRequestClose={onCancel}
      closeTimeoutMS={200}
      appElement={document.getElementById('root')}
    >
      <form onSubmit={onSubmit}>
        <input
          type="text"
          placeholder="Group name (required)"
          value={groupName}
          onChange={onGroupNameChange}
          autoFocus
        />
        <input
          type="text"
          placeholder="Module code/Class name (Optional)"
          value={module}
          onChange={onModuleChange}
        />
        <button>Submit</button>
        <button onClick={onCancel}>Cancel</button>
        <div>{error}</div>
      </form>
    </Modal>
  );
};

const mapDispatchToProps = dispatch => ({
  startNewGroup: (groupName, module) =>
    dispatch(startNewGroup(groupName, module))
});

export default connect(null, mapDispatchToProps)(AddGroupModal);
