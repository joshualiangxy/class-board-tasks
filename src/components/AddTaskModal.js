import React from 'react';
import Modal from 'react-modal';
import { connect } from 'react-redux';
import { startAddPersonalTask, startAddGroupTask } from '../actions/tasks';
import TaskForm from './TaskForm';

export const AddTaskModal = ({
  startAddPersonalTask,
  startAddGroupTask,
  removeAdmin,
  isOpen,
  onRequestClose,
  gid,
  groupModule,
  groupName,
  addGroupTask,
  users
}) => {
  const submitTask = task =>
    gid
      ? startAddGroupTask(task, groupName)
          .then(isAdmin => {
            if (isAdmin) return addGroupTask(task);
            else return removeAdmin();
          })
          .then(() => onRequestClose())
      : startAddPersonalTask(task).then(() => onRequestClose());

  return (
    <Modal
      className="task-modal"
      isOpen={isOpen}
      contentLabel="Add Task"
      onRequestClose={onRequestClose}
      closeTimeoutMS={200}
      appElement={document.getElementById('root')}
    >
      <h1>Add Task</h1>
      <TaskForm
        gid={gid}
        groupModule={groupModule}
        users={users}
        submitTask={submitTask}
        onRequestClose={onRequestClose}
      />
    </Modal>
  );
};

const mapDispatchToProps = dispatch => ({
  startAddPersonalTask: task => dispatch(startAddPersonalTask(task)),
  startAddGroupTask: (task, groupName) =>
    dispatch(startAddGroupTask(task, groupName))
});

export default connect(undefined, mapDispatchToProps)(AddTaskModal);
