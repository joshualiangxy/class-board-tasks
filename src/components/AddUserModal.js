import React, { useState } from 'react';
import Modal from 'react-modal';
import firebase from '../firebase/firebase';

// Note: This still works off uid for now. This also does NOT work yet.
const AddUserModal = ({ isOpen, onRequestClose, group }) => {
  const addNewUser = (userEmail, group) => {
    const uid = userEmail; // TODO: Change this when email-uid are linked!

    // Both of these references are DocumentReference
    const userRef = firebase.firestore().collection('users').doc(uid);
    const groupRef = firebase.firestore().collection('groups').doc(group);

    // Gets the data from a user.
    userRef.get().then(doc => {
      const data = doc.data();
      // Write displayName, studentNum, and UID into a new document
      if (typeof data === 'undefined') {
        // Most likely a user that doesn't exist.
        setError('No such user found');
      } else {
        groupRef
          .collection('users')
          .doc(uid) // Create document for this user's UID
          .set({
            studentNum: data.studentNum,
            displayName: data.displayName,
            uid: uid,
            admin: false
          })
          .then(() => {
            setError('User added!');
          })
          .catch(error => {
            setError('Something went wrong.');
          });
      }
      // Defaults to non-admin.
    });
  };

  // TODO: When calling this modal, pass in the group UID as a prop
  const [userEmail, setUserEmail] = useState('');
  const [error, setError] = useState('');
  // const [usersAdded, setUsersAdded] = useState([]); // For listing down people that have been added?

  const onUserEmailChange = e => {
    setUserEmail(e.target.value);
  };

  const onCancel = () => {
    setUserEmail('');
    onRequestClose();
  };

  const onSubmit = e => {
    e.preventDefault();
    const submittedEmail = userEmail.trim();
    if (!submittedEmail) {
      setError('Please enter a group name');
    } else {
      addNewUser(submittedEmail, group);
      // TODO: Catch user not found
      // Does not close after submitting, so that the user can continue to add more users

      // setUserEmail(''); // Also resets the field.
      // setError('Added!');
    }
  };

  return (
    <Modal
      isOpen={isOpen}
      contentLabel="New Group"
      onRequestClose={() => onRequestClose()}
      appElement={document.getElementById('root')}
    >
      <form onSubmit={onSubmit}>
        <input
          type="text"
          placeholder="User ID (required)"
          value={userEmail}
          onChange={onUserEmailChange}
          autoFocus
        />
        <button>Submit</button>
        <button onClick={onCancel}>Cancel</button>
        <div>{error}</div>
      </form>
    </Modal>
  );
};

export default AddUserModal;
