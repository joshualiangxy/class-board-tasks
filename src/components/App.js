import React from 'react';
import '../styles/App.css';

const App = () => (
  <div>
    <h1>This is my app</h1>
  </div>
);

export default App;

