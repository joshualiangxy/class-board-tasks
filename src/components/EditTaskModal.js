import React from 'react';
import Modal from 'react-modal';
import { connect } from 'react-redux';
import {
  startEditPersonalTask,
  startEditGroupTask,
  startSetTasks
} from '../actions/tasks';
import TaskForm from './TaskForm';

export const EditTaskModal = ({
  startEditPersonalTask,
  startEditGroupTask,
  startSetTasks,
  isOpen,
  onRequestClose,
  gid,
  groupModule,
  groupName,
  task,
  users,
  editGroupTask
}) => {
  const submitTask = updates =>
    gid
      ? startEditGroupTask(task.id, updates, groupName).then(() => {
          editGroupTask(task.id, updates);
          startSetTasks(); // Temporary solution
          onRequestClose();
        })
      : startEditPersonalTask(task.id, updates).then(() => onRequestClose());

  return (
    <Modal
      isOpen={isOpen}
      contentLabel="Edit Task"
      onRequestClose={onRequestClose}
      appElement={document.getElementById('root')}
    >
      <h2>Edit Task</h2>
      <TaskForm
        gid={gid}
        groupModule={groupModule}
        users={users}
        submitTask={submitTask}
        onRequestClose={onRequestClose}
        task={task}
      />
    </Modal>
  );
};

const mapDispatchToProps = dispatch => ({
  startEditPersonalTask: (id, updates) =>
    dispatch(startEditPersonalTask(id, updates)),
  startEditGroupTask: (id, updates, groupName) =>
    dispatch(startEditGroupTask(id, updates, groupName)),
  startSetTasks: () => dispatch(startSetTasks())
});

export default connect(undefined, mapDispatchToProps)(EditTaskModal);
