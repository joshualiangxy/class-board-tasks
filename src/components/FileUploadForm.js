import React, { useState } from 'react';
import firebase from '../firebase/firebase';
import { startUpdateDownloadURL } from '../actions/tasks';
import { uploadFile, startRemoveUserFile } from '../actions/files';
import { connect } from 'react-redux';

export const FileUploadForm = ({
  id,
  uid,
  uploadFile,
  TaskEvent,
  startUpdateDownloadURL,
  startRemoveUserFile,
  fileExists
}) => {
  const [error, setError] = useState('');
  const [uploading, setUploading] = useState(false);

  const fileUpload = e => {
    e.preventDefault();
    e.persist();

    const element = document.getElementById(id);
    const files = element.files;
    if (files.length > 0) {
      const file = files[0];

      Promise.resolve()
        .then(() => {
          if (fileExists) return startRemoveUserFile(uid);
        })
        .then(() => {
          const uploadTask = uploadFile(file);

          uploadTask.on(
            TaskEvent.STATE_CHANGED,
            () => setUploading(true),
            error => {
              switch (error.code) {
                case 'storage/unauthorized':
                  setError('Not allowed to upload file!');
                  break;
                case 'storage/canceled':
                  setError('File upload cancelled');
                  break;
                default:
                  setError('File upload failed');
                  break;
              }
              setUploading(false);
            },
            () => {
              const ref = uploadTask.snapshot.ref;
              const fileName = ref.name;

              ref
                .getDownloadURL()
                .then(downloadURL => {
                  startUpdateDownloadURL(downloadURL, fileName);
                })
                .then(() => {
                  setError('');
                  setUploading(false);
                  e.target.reset();
                });
            }
          );
        });
    } else setError('No file selected');
  };

  return (
    <form onSubmit={fileUpload}>
      {error && <p>{error}</p>}
      <input
        type="file"
        id={id}
        onClick={e => e.stopPropagation()}
        accept=".pdf"
        disabled={uploading}
      />
      <button onClick={e => e.stopPropagation()} disabled={uploading}>
        {uploading ? 'Uploading...' : 'Upload'}
      </button>
    </form>
  );
};

const mapStateToProps = ({ tasks, auth }, { id }) => {
  const uid = auth.user.uid;
  const task = tasks.find(task => task.id === id);

  return {
    TaskEvent: firebase.storage.TaskEvent,
    fileExists: task ? task.downloadURLs.hasOwnProperty(uid) : false,
    uid
  };
};

const mapDispatchToProps = (dispatch, { id, gid, namingConvention }) => ({
  uploadFile: file => dispatch(uploadFile(file, id, namingConvention)),
  startUpdateDownloadURL: (downloadURL, fileName) =>
    dispatch(startUpdateDownloadURL(id, gid, downloadURL, fileName)),
  startRemoveUserFile: uid => dispatch(startRemoveUserFile(id, uid))
});

export default connect(mapStateToProps, mapDispatchToProps)(FileUploadForm);
