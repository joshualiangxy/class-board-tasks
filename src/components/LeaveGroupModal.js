import React, { useState, useEffect } from 'react';
import Modal from 'react-modal';
import { connect } from 'react-redux';
import { history } from '../routers/AppRouter';
import { startLeaveGroup } from '../actions/groups';

const LeaveGroupModal = ({
  isOpen,
  onRequestClose,
  startLeaveGroup,
  gid,
  users,
  renderLoad,
  admin,
  uid
}) => {
  const [canLeave, setCanLeave] = useState(false);

  // For determining if a user should be allowed to leave
  useEffect(() => {
    if (users.length <= 1) {
      // This user is the only one in the group, so let him leave.
      setCanLeave(true);
    } else if (users.find(user => user.uid !== uid && user.admin)) {
      // There is another user in this group who is an admin.
      setCanLeave(true);
    } else if (!admin) {
      // This person is not an admin.
      setCanLeave(true);
    } else {
      // This person is an admin, there are other people in the group, and there aren't any other admins.
      setCanLeave(false); // Can't leave.
    }
  }, [users, admin, uid]);

  // Passing users into this modal so that we can tell if we should clean up this (empty) group or not
  const onCancel = () => {
    onRequestClose();
  };

  const onSubmit = e => {
    e.preventDefault();
    // There is another admin in the group already, or this group will be empty after leaving.
    renderLoad();
    history.push('/groups');
    startLeaveGroup(gid, users.length).then(() => onRequestClose());
  };

  return (
    <Modal
      isOpen={isOpen}
      contentLabel="Leave Group"
      onRequestClose={() => onRequestClose()}
      appElement={document.getElementById('root')}
    >
      {canLeave && (
        <div>
          <h3>Do you want to leave?</h3>
          <button onClick={onSubmit}>Yes</button>
          <button onClick={onCancel}>No</button>
        </div>
      )}
      {!canLeave && (
        <div>
          <h3>You cannot leave this group.</h3>
          <button onClick={onCancel}>Return</button>
        </div>
      )}
    </Modal>
  );
};

const mapDispatchToProps = dispatch => ({
  startLeaveGroup: (gid, count) => dispatch(startLeaveGroup(gid, count))
});

export default connect(null, mapDispatchToProps)(LeaveGroupModal);
