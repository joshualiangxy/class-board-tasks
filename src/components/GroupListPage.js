import React, { useState } from 'react';
import AddGroupModal from './AddGroupModal';
import { connect } from 'react-redux';
import { Link } from 'react-router-dom';

export const GroupListPage = ({ groups }) => {
  const [open, setOpen] = useState(false);

  const onRequestClose = () => {
    setOpen(false);
  };

  const openNewGroup = () => {
    setOpen(true);
  };

  return (
    <div>
      <div className="page-header">
        <div className="content-container">
          <h1 className="page-header__title">Groups</h1>
          <button
            className="button button--norm page-header__actions"
            onClick={openNewGroup}
          >
            Add New Group
          </button>
        </div>
      </div>
      <div className="content-container">
        {
          // Add a 'no groups' when there's no groups
          groups.length === 0 ? (
            <p className="list-header__text">No groups</p>
          ) : (
            groups.map(group => (
              <div key={group.gid} className="grouplist">
                <Link
                  className="button--link-black"
                  to={`/groups/${group.gid}`}
                >
                  <h2 className="groupitem">{group.name}</h2>
                  
                </Link>
              </div>
            ))
          )
        }
      </div>

      <AddGroupModal isOpen={open} onRequestClose={onRequestClose} />
    </div>
  );
};

const mapStateToProps = state => ({ groups: state.groups });

export default connect(mapStateToProps)(GroupListPage);
