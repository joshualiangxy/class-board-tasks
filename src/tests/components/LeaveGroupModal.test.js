import React from 'react';
import { shallow } from 'enzyme';
import { LeaveGroupModal } from '../../components/LeaveGroupModal';
import groupUsers from '../fixtures/groupUsers';

const isOpen = true;
const onRequestClose = jest.fn();
const startLeaveGroup = jest.fn();
const gid = '72b6b3bb-cb74-4d08-ae67-dfc1b51baaf9';
const users = groupUsers;
const renderLoad = jest.fn();
const admin = false;
const uid = 'testuid2';

const wrapper = shallow(
  <LeaveGroupModal
    isOpen={isOpen}
    onRequestClose={onRequestClose}
    startLeaveGroup={startLeaveGroup}
    gid={gid}
    users={users}
    renderLoad={renderLoad}
    admin={admin}
    uid={uid}
  />
);

describe('render', () => {
  it('should allow the non-admin user to leave', () => {
    expect(wrapper).toMatchSnapshot();
  });
  it('should not allow the admin user to leave', () => {
    wrapper.setProps({ uid: 'testuid', admin: true });
    expect(wrapper).toMatchSnapshot();
  });
  it('should allow the only user left to leave', () => {
    wrapper.setProps({ users: users.filter(u => u.uid === 'testuid') });
    expect(wrapper).toMatchSnapshot();
  });
});

describe('leave', () => {
  it('should call startLeaveGroup', () => {
    wrapper.setProps({ uid: 'testuid2', admin: false, users });
    wrapper
      .find('button')
      .at(0)
      .prop('onClick')({ preventDefault: jest.fn() })
      .then(() => {
        expect(startLeaveGroup).toHaveBeenCalledTimes(1);
        expect(startLeaveGroup).toHaveBeenLastCalledWith(gid, 2);
      });
  });
});
