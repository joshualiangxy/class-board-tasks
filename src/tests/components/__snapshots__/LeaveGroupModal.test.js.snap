// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`render should allow the non-admin user to leave 1`] = `
<Modal
  appElement={null}
  ariaHideApp={true}
  bodyOpenClassName="ReactModal__Body--open"
  className="setting-modal"
  closeTimeoutMS={200}
  contentLabel="Leave Group"
  isOpen={true}
  onRequestClose={[Function]}
  parentSelector={[Function]}
  portalClassName="ReactModalPortal"
  role="dialog"
  shouldCloseOnEsc={true}
  shouldCloseOnOverlayClick={true}
  shouldFocusAfterRender={true}
  shouldReturnFocusAfterClose={true}
>
  <div>
    <h3>
      Do you want to leave?
    </h3>
    <button
      className="button button--norm button--right"
      onClick={[Function]}
    >
      Yes
    </button>
    <button
      className="button button--norm button--right"
      onClick={[Function]}
    >
      No
    </button>
  </div>
</Modal>
`;

exports[`render should allow the only user left to leave 1`] = `
<Modal
  appElement={null}
  ariaHideApp={true}
  bodyOpenClassName="ReactModal__Body--open"
  className="setting-modal"
  closeTimeoutMS={200}
  contentLabel="Leave Group"
  isOpen={true}
  onRequestClose={[Function]}
  parentSelector={[Function]}
  portalClassName="ReactModalPortal"
  role="dialog"
  shouldCloseOnEsc={true}
  shouldCloseOnOverlayClick={true}
  shouldFocusAfterRender={true}
  shouldReturnFocusAfterClose={true}
>
  <div>
    <h3>
      Do you want to leave?
    </h3>
    <button
      className="button button--norm button--right"
      onClick={[Function]}
    >
      Yes
    </button>
    <button
      className="button button--norm button--right"
      onClick={[Function]}
    >
      No
    </button>
  </div>
</Modal>
`;

exports[`render should not allow the admin user to leave 1`] = `
<Modal
  appElement={null}
  ariaHideApp={true}
  bodyOpenClassName="ReactModal__Body--open"
  className="setting-modal"
  closeTimeoutMS={200}
  contentLabel="Leave Group"
  isOpen={true}
  onRequestClose={[Function]}
  parentSelector={[Function]}
  portalClassName="ReactModalPortal"
  role="dialog"
  shouldCloseOnEsc={true}
  shouldCloseOnOverlayClick={true}
  shouldFocusAfterRender={true}
  shouldReturnFocusAfterClose={true}
>
  <div>
    <h3>
      You cannot leave this group as the only admin!
    </h3>
    <button
      className="button button--norm button--right"
      onClick={[Function]}
    >
      Return
    </button>
  </div>
</Modal>
`;
